README.md
# NgxGallery

A simple native gallery component for Angular 10+.

- Show thumbnail in preview mode.
- Remove BrowserAnimationsModule from gallery package.

## Getting Started

### Installation

Install via [npm][npm] package manager

```bash
npm install @artavil/ngx-gallery --save
```

### Versions

1.x.x and above - for Angular v10+

### Usage

Import `ngx-gallery` module

```js
import { HttpClientModule} from '@angular/common/http';
import { NgxGalleryModule } from '@artavil/ngx-gallery';

@NgModule({
  imports: [ HttpClientModule, NgxGalleryModule ]
})
```

Then in HTML

```html
<ngx-gallery
  [options]="galleryOptions"
  [images]="galleryImages"
  class="ngx-gallery"
></ngx-gallery>
```

where

```
import {Component, OnInit} from '@angular/core';
import {NgxGalleryOptions} from '@artavil/ngx-gallery';
import {NgxGalleryImage} from '@artavil/ngx-gallery';
import {NgxGalleryAnimation} from '@artavil/ngx-gallery';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent implements OnInit{
  galleryOptions: NgxGalleryOptions[];
  galleryImages: NgxGalleryImage[];

  constructor() { }

  ngOnInit() {
    this.galleryOptions = [
      {
        width: '600px',
        height: '400px',
        thumbnailsColumns: 4,
        imageAnimation: NgxGalleryAnimation.Slide,
        previewThumbnail: true
      },
      // max-width 800
      {
        breakpoint: 800,
        width: '100%',
        height: '600px',
        imagePercent: 80,
        thumbnailsPercent: 20,
        thumbnailsMargin: 20,
        thumbnailMargin: 20
      },
      // max-width 400
      {
        breakpoint: 400,
        preview: false
      }
    ];

    this.galleryImages = [
      {
        small: 'assets/img/gallery/1-small.jpeg',
        medium: 'assets/img/gallery/1-medium.jpeg',
        big: 'assets/img/gallery/1-big.jpeg'
      },
      {
        small: 'assets/img/gallery/2-small.jpeg',
        medium: 'assets/img/gallery/2-medium.jpeg',
        big: 'assets/img/gallery/2-big.jpeg'
      },
      {
        small: 'assets/img/gallery/3-small.jpeg',
        medium: 'assets/img/gallery/3-medium.jpeg',
        big: 'assets/img/gallery/3-big.jpeg'
      },{
        small: 'assets/img/gallery/4-small.jpeg',
        medium: 'assets/img/gallery/4-medium.jpeg',
        big: 'assets/img/gallery/4-big.jpeg'
      },
      {
        small: 'assets/img/gallery/5-small.jpeg',
        medium: 'assets/img/gallery/5-medium.jpeg',
        big: 'assets/img/gallery/5-big.jpeg'
      }
    ];
  }
}
```

add this class to app CSS `.ngx-gallery { display: inline-block; margin-bottom: 20px; }`

## API

### Inputs

| Input     | Type                  | Default | Required | Description                    |
| --------- | --------------------- | ------- | -------- | ------------------------------ |
| [options] | `NgxGalleryOptions[]` | `-`     | no       | Config options for the Gallery |
| [images]  | `NgxGalleryImage[]`   | `-`     | no       | Images array                   |

### Outputs

| Output          | Description                       |
| --------------- | --------------------------------- |
| (change)        | Triggered on image change         |
| (imagesReady)   | Triggered when images length > 0  |
| (previewOpen)   | Triggered on preview open         |
| (previewClose)  | Triggered on preview close        |
| (previewChange) | Triggered on preview image change |

### Methods

| Name                              | Description                                      |
| --------------------------------- | ------------------------------------------------ |
| show(index: number): void         | Shows image at index                             |
| showNext(): void                  | Shows next image                                 |
| showPrev(): void                  | Shows prev image                                 |
| canShowNext(): boolean            | Returns true if there is next image              |
| canShowPrev(): boolean            | Returns true if there is prev image              |
| openPreview(index: number): void  | Opens preview at index                           |
| moveThumbnailsLeft(): void        | Moves thumbnails to left                         |
| moveThumbnailsRight(): void       | Moves thumbnails to right                        |
| canMoveThumbnailsLeft(): boolean  | Returns true if you can move thumbnails to left  |
| canMoveThumbnailsRight(): boolean | Returns true if you can move thumbnails to right |

### NgxGalleryOptions

#### Layout options

| Input       | Type      | Default                   | Required | Description                                             |
| ----------- | --------- | ------------------------- | -------- | ------------------------------------------------------- |
| width       | `string`  | `500px`                   | no       | Gallery width                                           |
| height      | `string`  | `400px`                   | no       | Gallery height                                          |
| breakpoint  | `number`  | `undefined`               | no       | Responsive breakpoint, works like media query max-width |
| fullWidth   | `bolean`  | `false`                   | no       | Sets the same width as browser                          |
| layout      | `string`  | `NgxGalleryLayout.Bottom` | no       | Sets thumbnails position                                |
| startIndex  | `number`  | `0`                       | no       | Sets index of selected image on start                   |
| linkTarget  | `string`  | `_blank`                  | no       | Sets target attribute of link                           |
| lazyLoading | `boolean` | `true`                    | no       | Enables/disables lazy loading for images                |

#### Image options

| Input                     | Type                 | Default                     | Required | Description                                  |
| ------------------------- | -------------------- | --------------------------- | -------- | -------------------------------------------- |
| image                     | `boolean`            | `true`                      | no       | Enables or disables image                    |
| imageDescription          | `boolean`            | `true`                      | no       | Enables or disables description for images   |
| imagePercent              | `number`             | `75`                        | no       | Percentage height                            |
| imageArrows               | `boolean`            | `true`                      | no       | Enables or disables arrows                   |
| imageArrowsAutoHide       | `boolean`            | `false`                     | no       | Enables or disables arrows auto hide         |
| imageSwipe                | `boolean`            | `false`                     | no       | Enables or disables swipe                    |
| imageAnimation            | `string`             | `NgxGalleryAnimation.Fade`  | no       | Animation type                               |
| imageSize                 | `string`             | `NgxGalleryImageSize.Cover` | no       | Image size                                   |
| imageAutoPlay             | `boolean`            | `false`                     | no       | Enables or disables auto play                |
| imageAutoPlayInterval     | `number`             | `2000`                      | no       | Interval for auto play (ms)                  |
| imageAutoPlayPauseOnHover | `boolean`            | `false`                     | no       | Enables or disables pause auto play on hover |
| imageInfinityMove         | `boolean`            | `false`                     | no       | Enables or disables infinity move by arrows  |
| imageActions              | `NgxGalleryAction[]` | `[]`                        | no       | Enables or disables navigation bullets       |

#### Thumbnails options

| Input                    | Type                 | Default                     | Required | Description                                                              |
| ------------------------ | -------------------- | --------------------------- | -------- | ------------------------------------------------------------------------ |
| thumbnails               | `boolean`            | `true`                      | no       | Enables or disables thumbnails                                           |
| thumbnailsColumns        | `number`             | `4`                         | no       | Columns count                                                            |
| thumbnailsRows           | `number`             | `1`                         | no       | Rows count                                                               |
| thumbnailsPercent        | `number`             | `25`                        | no       | Percentage height                                                        |
| thumbnailsMargin         | `number`             | `10`                        | no       | Margin between thumbnails and image                                      |
| thumbnailsArrows         | `boolean`            | `true`                      | no       | Enables or disables arrows                                               |
| thumbnailsArrowsAutoHide | `boolean`            | `false`                     | no       | Enables or disables arrows auto hide                                     |
| thumbnailsSwipe          | `boolean`            | `false`                     | no       | Enables or disables swipe                                                |
| thumbnailsMoveSize       | `number`             | `1`                         | no       | Number of items to move on arrow click                                   |
| thumbnailsOrder          | `number`             | `NgxGalleryOrder.Column`    | no       | Images order                                                             |
| thumbnailsRemainingCount | `boolean`            | `false`                     | no       | If true arrows are disabled and last item has label with remaining count |
| thumbnailsAsLinks        | `boolean`            | `false`                     | no       | Enables or disables links on thumbnails                                  |
| thumbnailsAutoHide       | `boolean`            | `false`                     | no       | Hides thumbnails if there is only one image                              |
| thumbnailMargin          | `number`             | `10`                        | no       | Margin between images in thumbnails                                      |
| thumbnailSize            | `string`             | `NgxGalleryImageSize.Cover` | no       | Thumbnail size                                                           |
| thumbnailActions         | `NgxGalleryAction[]` | `[]`                        | no       | Array of custom actions                                                  |
| thumbnailClasses         | `string[]`           | `[]`                        | no       | Custom classes to add to thumbnail component                             |

#### Preview options

| Input                       | Type              | Default | Required | Description                                            |
| --------------------------- | ----------------- | ------- | -------- | ------------------------------------------------------ |
| preview                     | `boolean`         | `true`  | no       | Enables or disables preview                            |
| previewDescription          | `boolean`         | `true`  | no       | Enables or disables description for images             |
| previewArrows               | `boolean`         | `true`  | no       | Enables or disables arrows                             |
| previewArrowsAutoHide       | boolean: `string` | `false` | no       | Enables or disables arrows auto hide                   |
| previewSwipe                | `boolean`         | `false` | no       | Enables or disables swipe                              |
| previewFullscreen           | `boolean`         | `false` | no       | Enables or disables fullscreen icon                    |
| previewForceFullscreen      | `boolean`         | `false` | no       | Enables or disables opening preview in fullscreen mode |
| previewCloseOnClick         | `boolean`         | `false` | no       | Enables or disables closing preview by click           |
| previewCloseOnEsc           | `boolean`         | `false` | no       | Enables or disables closing preview by esc keyboard    |
| previewKeyboardNavigation   | `boolean`         | `false` | no       | Enables or disables navigation by keyboard             |
| previewAnimation            | `boolean`         | `true`  | no       | Enables or disables image loading animation            |
| previewAutoPlay             | `boolean`         | `false` | no       | Enables or disables auto play                          |
| previewAutoPlayInterval     | `number`          | `2000`  | no       | Interval for auto play (ms)                            |
| previewAutoPlayPauseOnHover | `boolean`         | `false` | no       | Enables or disables pouse auto play on hover           |
| previewInfinityMove         | `boolean`         | `false` | no       | Enables or disables infinity move by arrows            |
| previewZoom                 | `boolean`         | `false` | no       | Enables or disables zoom in and zoom out               |
| previewZoomStep             | `number`          | `0.1`   | no       | Step for zoom change                                   |
| previewZoomMax              | `number`          | `2`     | no       | Max value for zoom                                     |
| previewZoomMin              | `number`          | `0.5`   | no       | Min value for zoom                                     |
| previewRotate               | `boolean`         | `false` | no       | Enables or disables rotate buttons                     |
| previewDownload             | `boolean`         | `false` | no       | Enables or disables download button                    |
| previewBullets              | `boolean`         | `false` | no       | Enables or disables navigation bullets                 |
| previewThumbnail            | `boolean`         | `false` | no       | Enables or disables thumbnail in preview               |

#### Icons options

| Input           | Type                 | Default                                | Required | Description                                                                                           |
| --------------- | -------------------- | -------------------------------------- | -------- | ----------------------------------------------------------------------------------------------------- |
| arrowPrevIcon   | `string`             | `'fa fa-arrow-circle-left'`            | no       | Icon for prev arrow                                                                                   |
| arrowNextIcon   | `string`             | `'fa fa-arrow-circle-right'`           | no       | Icon for next arrow                                                                                   |
| closeIcon       | `string`             | `'fa fa-times-circle'`                 | no       | Icon for close button                                                                                 |
| fullscreenIcon  | `string`             | `'fa fa-arrows-alt'`                   | no       | Icon for fullscreen button                                                                            |
| spinnerIcon     | `string`             | `'fa fa-spinner fa-pulse fa-3x fa-fw'` | no       | Icon for spinner                                                                                      |
| zoomInIcon      | `string`             | `'fa fa-search-plus'`                  | no       | Icon for zoom in                                                                                      |
| zoomOutIcon     | `string`             | `'fa fa-search-minus'`                 | no       | Icon for zoom out                                                                                     |
| rotateLeftIcon  | `string`             | `'fa fa-undo'`                         | no       | Icon for rotate left                                                                                  |
| rotateRightIcon | `string`             | `'fa fa-repeat'`                       | no       | Icon for rotate right                                                                                 |
| downloadIcon    | `string`             | `'fa fa-arrow-circle-down'`            | no       | Icon for download                                                                                     |
| actions         | `NgxGalleryAction[]` | `[]`                                   | no       | Array of new custom actions that will be added to the left of the current close/zoom/fullscreen icons |

### NgxGalleryImage

| Input       | Type                     | Default | Required | Description                                        |
| ----------- | ------------------------ | ------- | -------- | -------------------------------------------------- |
| small       | `string/SafeResourceUrl` | `-`     | no       | Url used in thumbnails                             |
| medium      | `string/SafeResourceUrl` | `-`     | no       | Url used in image                                  |
| big         | `string/SafeResourceUrl` | `-`     | yes      | Url used in preview                                |
| description | `string`                 | `-`     | no       | Description used in preview                        |
| url         | `string`                 | `-`     | no       | Url used in link                                   |
| label       | `string`                 | `-`     | no       | Label used for aria-label when thumbnail is a link |

### NgxGalleryAnimation

- `Fade` (default)
- `Slide`
- `Rotate`
- `Zoom`

### NgxGalleryImageSize

- `Cover` (default)
- `Contain`

### NgxGalleryLayout

- `Top`
- `Bottom` (default)

### NgxGalleryOrder

- `Column` (default)
- `Row`
- `Page`

### NgxGalleryAction

- `icon` | Type: `string` - icon for custom action
- `disabled` | Type: `boolean` | Default value: `false` - if the icon should be disabled
- `titleText` | Type: `string` | Default value: `''` - text to set the title attribute to
- `onClick` | Type: `(event: Event, index: number) => void` - Output function to call when custom action icon is clicked

## What's included

Within the download you'll find the following directories and files. You'll see something like this:

```
ngx-gallery/
└── projects/
    ├── gallery/
    └── gallery-app/
```

`gallery/` - library

`gallery-app/` - demo application

## Contributing

Please read through our [contributing guidelines](https://github.com/AhmadPirzargar/ngx-gallery/blob/master/CONTRIBUTING.md). Included are directions for opening issues, coding standards, and notes on development.

Editor preferences are available in the [editor config](https://github.com/AhmadPirzargar/ngx-gallery/blob/master/.editorconfig) for easy use in common text editors. Read more and download plugins at <http://editorconfig.org>.

## Versioning

For transparency into our release cycle and in striving to maintain backward compatibility, NgxGallery is maintained under [the Semantic Versioning guidelines](http://semver.org/).

See [the Releases section of our project](https://github.com/AhmadPirzargar/ngx-gallery/releases) for changelogs for each release version.

## Creators

**Ahmad Pirzargar**

- <https://github.com/AhmadPirzargar>

## Credits

This library is being fully rewritten for next Angular versions from original abandoned library written by Łukasz Gałka. I maintained full compatibility with the original library at the api level.
<https://github.com/lukasz-galka/ngx-gallery>


CHANGELOG.md
<a name="1.1.0"></a>

# 1.1.0 (2021-10-26)

- Preview thumbnail added

<a name="1.0.11"></a>

# 1.0.11 (2020-03-05)

- Need to add ngx-gallery class to CSS documentation #8
- Unable to get custom icons to work #6
- Mention about original lib #7

<a name="1.0.0"></a>

# 1.0.0 (2019-11-24)

- Initial release


projects/gallery/src/lib/ngx-gallery.module.ts
import { NgModule, Injectable } from '@angular/core';
import { NgxGalleryComponent } from './ngx-gallery.component';
import { NgxGalleryImageComponent } from './ngx-gallery-image/ngx-gallery-image.component';
import { NgxGalleryArrowsComponent } from './ngx-gallery-arrows/ngx-gallery-arrows.component';
import { NgxGalleryThumbnailsComponent } from './ngx-gallery-thumbnails/ngx-gallery-thumbnails.component';
import { NgxGalleryPreviewComponent } from './ngx-gallery-preview/ngx-gallery-preview.component';
import { NgxGalleryActionComponent } from './ngx-gallery-action/ngx-gallery-action.component';
import { NgxGalleryBulletsComponent } from './ngx-gallery-bullets/ngx-gallery-bullets.component';
import { CommonModule } from '@angular/common';
import { HAMMER_GESTURE_CONFIG, HammerGestureConfig } from '@angular/platform-browser';

@Injectable()
export class CustomHammerConfig extends HammerGestureConfig {
  overrides = {
    pinch: { enable: false },
    rotate: { enable: false }
  } as any;
}


@NgModule({
  declarations: [
    NgxGalleryComponent,
    NgxGalleryImageComponent,
    NgxGalleryArrowsComponent,
    NgxGalleryThumbnailsComponent,
    NgxGalleryPreviewComponent,
    NgxGalleryActionComponent,
    NgxGalleryBulletsComponent
  ],
  imports: [CommonModule],
  exports: [NgxGalleryComponent],
  providers: [
    { provide: HAMMER_GESTURE_CONFIG, useClass: CustomHammerConfig }
  ]
})
export class NgxGalleryModule {
}


projects/gallery/src/lib/ngx-gallery-options.ts
import { NgxGalleryAction } from './ngx-gallery-action';
import { NgxGalleryOrder } from './ngx-gallery-order';
import { NgxGalleryLayout } from './ngx-gallery-layout';
import { NgxGalleryAnimation } from './ngx-gallery-animation';
import { NgxGalleryImageSize } from './ngx-gallery-image-size';

export interface INgxGalleryOptions {
  width?: string;
  height?: string;
  breakpoint?: number;
  fullWidth?: boolean;
  layout?: string;
  startIndex?: number;
  linkTarget?: string;
  lazyLoading?: boolean;
  image?: boolean;
  imagePercent?: number;
  imageArrows?: boolean;
  imageArrowsAutoHide?: boolean;
  imageSwipe?: boolean;
  imageAnimation?: string;
  imageSize?: string;
  imageAutoPlay?: boolean;
  imageAutoPlayInterval?: number;
  imageAutoPlayPauseOnHover?: boolean;
  imageInfinityMove?: boolean;
  imageActions?: NgxGalleryAction[];
  imageDescription?: boolean;
  imageBullets?: boolean;
  thumbnails?: boolean;
  thumbnailsColumns?: number;
  thumbnailsRows?: number;
  thumbnailsPercent?: number;
  thumbnailsMargin?: number;
  thumbnailsArrows?: boolean;
  thumbnailsArrowsAutoHide?: boolean;
  thumbnailsSwipe?: boolean;
  thumbnailsMoveSize?: number;
  thumbnailsOrder?: NgxGalleryOrder;
  thumbnailsRemainingCount?: boolean;
  thumbnailsAsLinks?: boolean;
  thumbnailsAutoHide?: boolean;
  thumbnailMargin?: number;
  thumbnailSize?: string;
  thumbnailActions?: NgxGalleryAction[];
  thumbnailClasses?: string[];
  preview?: boolean;
  previewDescription?: boolean;
  previewArrows?: boolean;
  previewArrowsAutoHide?: boolean;
  previewSwipe?: boolean;
  previewFullscreen?: boolean;
  previewForceFullscreen?: boolean;
  previewCloseOnClick?: boolean;
  previewCloseOnEsc?: boolean;
  previewKeyboardNavigation?: boolean;
  previewAnimation?: boolean;
  previewAutoPlay?: boolean;
  previewAutoPlayInterval?: number;
  previewAutoPlayPauseOnHover?: boolean;
  previewInfinityMove?: boolean;
  previewZoom?: boolean;
  previewZoomStep?: number;
  previewZoomMax?: number;
  previewZoomMin?: number;
  previewRotate?: boolean;
  previewDownload?: boolean;
  previewThumbnail?: boolean;
  previewCustom?: (index: number) => void;
  previewBullets?: boolean;
  arrowPrevIcon?: string;
  arrowNextIcon?: string;
  closeIcon?: string;
  fullscreenIcon?: string;
  spinnerIcon?: string;
  zoomInIcon?: string;
  zoomOutIcon?: string;
  rotateLeftIcon?: string;
  rotateRightIcon?: string;
  downloadIcon?: string;
  actions?: NgxGalleryAction[];
}

export class NgxGalleryOptions implements INgxGalleryOptions {
  width?: string;
  height?: string;
  breakpoint?: number;
  fullWidth?: boolean;
  layout?: string;
  startIndex?: number;
  linkTarget?: string;
  lazyLoading?: boolean;
  image?: boolean;
  imagePercent?: number;
  imageArrows?: boolean;
  imageArrowsAutoHide?: boolean;
  imageSwipe?: boolean;
  imageAnimation?: string;
  imageSize?: string;
  imageAutoPlay?: boolean;
  imageAutoPlayInterval?: number;
  imageAutoPlayPauseOnHover?: boolean;
  imageInfinityMove?: boolean;
  imageActions?: NgxGalleryAction[];
  imageDescription?: boolean;
  imageBullets?: boolean;
  thumbnails?: boolean;
  thumbnailsColumns?: number;
  thumbnailsRows?: number;
  thumbnailsPercent?: number;
  thumbnailsMargin?: number;
  thumbnailsArrows?: boolean;
  thumbnailsArrowsAutoHide?: boolean;
  thumbnailsSwipe?: boolean;
  thumbnailsMoveSize?: number;
  thumbnailsOrder?: NgxGalleryOrder;
  thumbnailsRemainingCount?: boolean;
  thumbnailsAsLinks?: boolean;
  thumbnailsAutoHide?: boolean;
  thumbnailMargin?: number;
  thumbnailSize?: string;
  thumbnailActions?: NgxGalleryAction[];
  thumbnailClasses?: string[];
  preview?: boolean;
  previewDescription?: boolean;
  previewArrows?: boolean;
  previewArrowsAutoHide?: boolean;
  previewSwipe?: boolean;
  previewFullscreen?: boolean;
  previewForceFullscreen?: boolean;
  previewCloseOnClick?: boolean;
  previewCloseOnEsc?: boolean;
  previewKeyboardNavigation?: boolean;
  previewAnimation?: boolean;
  previewAutoPlay?: boolean;
  previewAutoPlayInterval?: number;
  previewAutoPlayPauseOnHover?: boolean;
  previewInfinityMove?: boolean;
  previewZoom?: boolean;
  previewZoomStep?: number;
  previewZoomMax?: number;
  previewZoomMin?: number;
  previewRotate?: boolean;
  previewDownload?: boolean;
  previewThumbnail?: boolean;
  previewCustom?: (index: number) => void;
  previewBullets?: boolean;
  arrowPrevIcon?: string;
  arrowNextIcon?: string;
  closeIcon?: string;
  fullscreenIcon?: string;
  spinnerIcon?: string;
  zoomInIcon?: string;
  zoomOutIcon?: string;
  rotateLeftIcon?: string;
  rotateRightIcon?: string;
  downloadIcon?: string;
  actions?: NgxGalleryAction[];

  constructor(obj: INgxGalleryOptions) {

    const preventDefaults = obj.breakpoint === undefined ? false : true;

    function use<T>(source: T, defaultValue: T): T {
      return obj && (source !== undefined || preventDefaults) ? source : defaultValue;
    }

    this.breakpoint = use(obj.breakpoint, undefined);
    this.width = use(obj.width, '500px');
    this.height = use(obj.height, '400px');
    this.fullWidth = use(obj.fullWidth, false);
    this.layout = use(obj.layout, NgxGalleryLayout.ThumbnailsBottom);
    this.startIndex = use(obj.startIndex, 0);
    this.linkTarget = use(obj.linkTarget, '_blank');
    this.lazyLoading = use(obj.lazyLoading, true);

    this.image = use(obj.image, true);
    this.imagePercent = use(obj.imagePercent, 75);
    this.imageArrows = use(obj.imageArrows, true);
    this.imageArrowsAutoHide = use(obj.imageArrowsAutoHide, false);
    this.imageSwipe = use(obj.imageSwipe, false);
    this.imageAnimation = use(obj.imageAnimation, NgxGalleryAnimation.Fade);
    this.imageSize = use(obj.imageSize, NgxGalleryImageSize.Cover);
    this.imageAutoPlay = use(obj.imageAutoPlay, false);
    this.imageAutoPlayInterval = use(obj.imageAutoPlayInterval, 2000);
    this.imageAutoPlayPauseOnHover = use(obj.imageAutoPlayPauseOnHover, false);
    this.imageInfinityMove = use(obj.imageInfinityMove, false);
    if (obj && obj.imageActions && obj.imageActions.length) {
      obj.imageActions = obj.imageActions.map(action => new NgxGalleryAction(action));
    }
    this.imageActions = use(obj.imageActions, []);
    this.imageDescription = use(obj.imageDescription, false);
    this.imageBullets = use(obj.imageBullets, false);

    this.thumbnails = use(obj.thumbnails, true);
    this.thumbnailsColumns = use(obj.thumbnailsColumns, 4);
    this.thumbnailsRows = use(obj.thumbnailsRows, 1);
    this.thumbnailsPercent = use(obj.thumbnailsPercent, 25);
    this.thumbnailsMargin = use(obj.thumbnailsMargin, 10);
    this.thumbnailsArrows = use(obj.thumbnailsArrows, true);
    this.thumbnailsArrowsAutoHide = use(obj.thumbnailsArrowsAutoHide, false);
    this.thumbnailsSwipe = use(obj.thumbnailsSwipe, false);
    this.thumbnailsMoveSize = use(obj.thumbnailsMoveSize, 1);
    this.thumbnailsOrder = use(obj.thumbnailsOrder, NgxGalleryOrder.Column);
    this.thumbnailsRemainingCount = use(obj.thumbnailsRemainingCount, false);
    this.thumbnailsAsLinks = use(obj.thumbnailsAsLinks, false);
    this.thumbnailsAutoHide = use(obj.thumbnailsAutoHide, false);
    this.thumbnailMargin = use(obj.thumbnailMargin, 10);
    this.thumbnailSize = use(obj.thumbnailSize, NgxGalleryImageSize.Cover);
    if (obj && obj.thumbnailActions && obj.thumbnailActions.length) {
      obj.thumbnailActions = obj.thumbnailActions.map(action => new NgxGalleryAction(action));
    }
    this.thumbnailActions = use(obj.thumbnailActions, []);
    this.thumbnailClasses = use(obj.thumbnailClasses, []);


    this.preview = use(obj.preview, true);
    this.previewDescription = use(obj.previewDescription, true);
    this.previewArrows = use(obj.previewArrows, true);
    this.previewArrowsAutoHide = use(obj.previewArrowsAutoHide, false);
    this.previewSwipe = use(obj.previewSwipe, false);
    this.previewFullscreen = use(obj.previewFullscreen, false);
    this.previewForceFullscreen = use(obj.previewForceFullscreen, false);
    this.previewCloseOnClick = use(obj.previewCloseOnClick, false);
    this.previewCloseOnEsc = use(obj.previewCloseOnEsc, false);
    this.previewKeyboardNavigation = use(obj.previewKeyboardNavigation, false);
    this.previewAnimation = use(obj.previewAnimation, true);
    this.previewAutoPlay = use(obj.previewAutoPlay, false);
    this.previewAutoPlayInterval = use(obj.previewAutoPlayInterval, 2000);
    this.previewAutoPlayPauseOnHover = use(obj.previewAutoPlayPauseOnHover, false);
    this.previewInfinityMove = use(obj.previewInfinityMove, false);
    this.previewZoom = use(obj.previewZoom, false);
    this.previewZoomStep = use(obj.previewZoomStep, 0.1);
    this.previewZoomMax = use(obj.previewZoomMax, 2);
    this.previewZoomMin = use(obj.previewZoomMin, 0.5);
    this.previewRotate = use(obj.previewRotate, false);
    this.previewDownload = use(obj.previewDownload, false);
    this.previewThumbnail = use(obj.previewThumbnail, false);
    this.previewCustom = use(obj.previewCustom, undefined);
    this.previewBullets = use(obj.previewBullets, false);

    this.arrowPrevIcon = use(obj.arrowPrevIcon, 'fa fa-arrow-circle-left');
    this.arrowNextIcon = use(obj.arrowNextIcon, 'fa fa-arrow-circle-right');
    this.closeIcon = use(obj.closeIcon, 'fa fa-times-circle');
    this.fullscreenIcon = use(obj.fullscreenIcon, 'fa fa-arrows-alt');
    this.spinnerIcon = use(obj.spinnerIcon, 'fa fa-spinner fa-pulse fa-3x fa-fw');
    this.zoomInIcon = use(obj.zoomInIcon, 'fa fa-search-plus');
    this.zoomOutIcon = use(obj.zoomOutIcon, 'fa fa-search-minus');
    this.rotateLeftIcon = use(obj.rotateLeftIcon, 'fa fa-undo');
    this.rotateRightIcon = use(obj.rotateRightIcon, 'fa fa-repeat');
    this.downloadIcon = use(obj.downloadIcon, 'fa fa-arrow-circle-down');

    if (obj && obj.actions && obj.actions.length) {
      obj.actions = obj.actions.map(action => new NgxGalleryAction(action));
    }
    this.actions = use(obj.actions, []);
  }
}


projects/gallery-app/src/app/app.component.ts
import { Component, OnInit } from '@angular/core';
import { NgxGalleryOptions } from '@artavil/ngx-gallery';
import { NgxGalleryImage } from '@artavil/ngx-gallery';
import { NgxGalleryAnimation } from '@artavil/ngx-gallery';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent implements OnInit {
  galleryOptions: NgxGalleryOptions[];
  galleryImages: NgxGalleryImage[];

  constructor() { }

  ngOnInit() {
    this.galleryOptions = [
      {
        width: '600px',
        height: '400px',
        thumbnailsColumns: 4,
        arrowPrevIcon: 'fa fa-chevron-left',
        arrowNextIcon: 'fa fa-chevron-right',
        imageAnimation: NgxGalleryAnimation.Rotate,
        imageSwipe: true,
        previewThumbnail: true
      },
      // max-width 800
      {
        breakpoint: 800,
        width: '100%',
        height: '600px',
        imagePercent: 80,
        thumbnailsPercent: 20,
        thumbnailsMargin: 20,
        thumbnailMargin: 20,
        imageSwipe: true
      },
      // max-width 400
      {
        breakpoint: 400,
        preview: false,
        imageSwipe: true
      }
    ];

    this.galleryImages = [
      {
        small: 'assets/img/gallery/1-small.jpeg',
        medium: 'assets/img/gallery/1-medium.jpeg',
        big: 'assets/img/gallery/1-big.jpeg'
      },
      {
        small: 'assets/img/gallery/2-small.jpeg',
        medium: 'assets/img/gallery/2-medium.jpeg',
        big: 'assets/img/gallery/2-big.jpeg'
      },
      {
        small: 'assets/img/gallery/3-small.jpeg',
        medium: 'assets/img/gallery/3-medium.jpeg',
        big: 'assets/img/gallery/3-big.jpeg'
      }, {
        small: 'assets/img/gallery/4-small.jpeg',
        medium: 'assets/img/gallery/4-medium.jpeg',
        big: 'assets/img/gallery/4-big.jpeg'
      },
      {
        small: 'https://html5box.com/html5gallery/images/BigBuckBunny_1.mp4',
        medium: 'https://html5box.com/html5gallery/images/BigBuckBunny_1.mp4',
        big: 'https://html5box.com/html5gallery/images/BigBuckBunny_1.mp4'
      }
    ];
  }
}
